src/ast/ast.h

```diff
@@ -27,7 +27,75 @@ class VisitorBase;
     return new T(*this);                                                       \
   };
 
-class Node {
+
+/* Update with
+
+   echo INVALID=0,; sed -n '/\%token$/,/;$/p' src/parser.yy | awk 'NF == 2 { print $1 "," }'
+
+ */
+enum class Token
```

Comment:
Should `Token` be in ast/ dir? Would it make more sense to have a parser/ dir with all the lexing and parsing files too?

Reply:
yeah maybe. Or we could limit the tokens to binop and unop symbols only to make ti less confusing

Reply:
From commit message: 

>  it reverses the relationship between the parser and ast

I think we can get rid of weird relationship ordering by just creating a `Tokens.h` file, right?

> Or we could limit the tokens to binop and unop symbols only to make ti less confusing

IMO that sounds more confusing. Probably better to keep it all in one place

Reply:
> From commit message:
> 
> > it reverses the relationship between the parser and ast
> 
> I think we can get rid of weird relationship ordering by just creating a `Tokens.h` file, right?

The parser would still use it's own tokens for the lexer->parser communication (also needs locations) so we can't get rid of that completely and still need to keep the list in 2 places which isn't great (not a real issue either)


> > Or we could limit the tokens to binop and unop symbols only to make ti less confusing
> 
> IMO that sounds more confusing. Probably better to keep it all in one place

I don't think it would be bad if we rename it from token to operator.

```
namespace ast {
   enum class Operator
     INCREMENT = 0,
     DECREMENT,
...
...
  }
}
```  

So in the parser it would be:

```
         |       DECREMENT map_or_var { $$ = new ast::Unop(ast::Operator::DECREMENT, $2, @1); } 
```

And usage in the ast

```
  switch (binop.op) {
    case Operator::EQ:
```

Reply:
> I don't think it would be bad if we rename it from token to operator.

I guess that makes sense. Just to be clear, this is a separate concern from build times now, right?

Reply:
Yes my goal is to reduce the build time after modifying the parser. Although not having to include parser.tab.hh in a lot of places is also a nice cleanup.